sympy/physics/mechanics/actuator.py

```diff
@@ -990,3 +991,154 @@ def __repr__(self):
         return (f"{self.__class__.__name__}("
                 f"{self.linear_stiffness}, {self.nonlinear_stiffness}, {self.pathway}, "
                 f"equilibrium_length={self.equilibrium_length})")
+
+class CoulombKineticFriction(ForceActuator):
+    """Coulomb kinetic friction with Stribeck and viscous effects.
+
+    Explanation
+    ===========
+
+    This represents a Coulomb kinetic friction with the Stribeck and viscous effect,
+    described by the function:
+
+    .. math::
+        F = f_c sign(v) + (f_{max} - f_c) e^{-(v / v_s)^2} + sigma v
+
+    where :math:`f_c` is the Coulomb friction constant, :math:`f_{max}` is the maximum static friction force,
+    :math:`v_s` is the coefficient of sliding friction, :math:`sigma` is the viscous friction constant,
+    and :math:`v` is the relative velocity.
+
+    The default friction force is :math:`F = mu_k N`, where :math:`N` is the normal force.
+    When specified, the actuator includes:
+
+    - Stribeck effect: :math:`(mu_s - mu_k) N e^{-(v / v_s)^2}`
+    - Viscous effect: :math:`sigma v`
+
+    In case we include all effects, the full actuator represents:
+
+    .. math::
+        F = mu_k N sign(v) + (mu_s - mu_k) N e^{-(v / v_s)^2} + sigma v
+
+    Please note that this actuator assumes slip and applies a force opposite to the velocity direction.
+
+    Parameters
+    ==========
+
+    normal_force : Expr
+        The normal force between the surfaces.
```

Comment:
A force is a vector, but you ask for a scalar expression here.

There is also no explanation of a sign convention for the measure number of this force.

Does this work when the force is not constant and changes sign?

Is the user expected to derive the normal force wrt to the pathway independently?

Reply:
I treated it as a scalar in the actuator class, test class, and simulation, which needs to be changed. And I think users need to independently derive the normal force according to their specific system, which I think needs to be mentioned as well.

Reply:
When finally calculating the force in the actuator, we have:
```
f_c = self.mu_k * self.normal_force
f_max = self.mu_s * self.normal_force
```

When we set `normal_force` as a vector, there's a conflict (simpify error) because `self.mu_k` and `self.mu_s` are sympy expressions, but `self.normal_force` is a vector.

Error:
```
...
E           sympy.core.sympify.SympifyError: SympifyError: "cannot sympify object of type <class 'sympy.physics.vector.vector.Vector'>"

../../../core/sympify.py:465: SympifyError
```

So I think using `.magnitude()` for `normal_force` can be a good candidate since it allows us to express `normal_force` as a vector while also enabling it to pass through the actuator without a conflict.

Reply:
Following the definition of this Coulomb actuator I would not expect the normal force to change sign (something for the user to check). Within the current definition of the equations, it is assumed that the normal force is defined as a positive scalar.
Changing it to a vector doesn't seem optimal to me. Firstly, the model is not valid when changing signs, so I would prefer not taking the Abs (I'd rather set the friction force to be zero if the normal force is negative). Secondly, introducing an abs (or sqrt(...**2)) is redundant.

Reply:
What if it is modeling the friction of a cylindrical joint? Couldn't that be valid regardless of the sign of the normal force?

Reply:
`.magnitude()` does not retain the information about the sign of the force.

Reply:
> What if it is modeling the friction of a cylindrical joint? Couldn't that be valid regardless of the sign of the normal force?

How would you define the normal force of a cylindrical joint as a vector? And how would you define the friction force? In several cases I would be inclined to define a torsional friction.

Reply:
A cylindrical joint that allows sliding and rotation. If a cylinder slides along a rod, then a force normal to the rod in any direction will cause friction.

Reply:
> A cylindrical joint that allows sliding and rotation. If a cylinder slides along a rod, then a force normal to the rod in any direction will cause friction.

Pathway is defined as a path between two points, called attachments, so I would say that these kind of cases are already constrained by the Pathway definition.

Reply:
A pathway has infinite normal vectors, so without more information one cannot fully define how friction might arise with an object engaging that pathway.

Reply:
Fair point, the direction of the normal force of a pathway is indeed undefined by default. However, as far as I can see there isn't really anything for which we would like to use the direction of the normal force in the `CoulombKineticActuator`.
If you prefer an option is to do an isinstance check and take the magnitude if it is a Vector and otherwise assume that it is a positive scalar.

Reply:
My preference is that it is clear to the user what this will and won't work with and what they have to do to make it work. Right now, things are a bit unclear except for the most simple case: block sliding on plane in linear path.

Reply:
Thanks for all the comments. I tried to model a cylindrical joint with Coulomb kinetic friction between the parent and child bodies. I modified `CoulombKineticFriction` to:
```
@property
def force(self):
    v = self.pathway.extension_velocity
    n_dir = self.normal_force.normalize()
    f_c = self.mu_k * self.normal_force.magnitude()
    f_max = self.mu_s * self.normal_force.magnitude()
    stribeck_term = (f_max - f_c) * n_dir * exp(-(v / self.v_s)**2)
    viscous_term = self.viscous_coefficient * v
    return f_c * sign(v) * n_dir + stribeck_term + viscous_term
```
so that it can treat the normal force as a vector. I used the code:
```
# A cylindrical joint with Coulomb kinetic friction between the parent and child bodies

# Define symbols
t = symbols('t')
q1 = dynamicsymbols('q1')  # Rotation coordinate
q2 = dynamicsymbols('q2')  # Translation coordinate
u1 = dynamicsymbols('u1')  # Rotation speed
u2 = dynamicsymbols('u2')  # Translation speed
mu_k = symbols('mu_k')  # Coefficient of kinetic friction
v_s = symbols('v_s')  # Coefficient of sliding friction
sigma = symbols('sigma')  # Viscous friction coefficient
r = symbols('r')  # Radius
N = symbols('N')  # Normal force magnitude

# Parent and child bodies
parent_frame = ReferenceFrame('P')
child_frame = ReferenceFrame('C')

O = Point('O')
parent_point = Point('parent_point')
child_point = Point('child_point')

# Position
parent_point.set_pos(O, r * parent_frame.y)  # Place parent_point on the surface of the cylinder
child_point.set_pos(parent_point, q2 * parent_frame.z)  # Initial position of child_point
child_frame.orient(parent_frame, 'axis', (q1, parent_frame.x))

parent = RigidBody('parent', parent_point, parent_frame, 1, (inertia(parent_frame, 1, 1, 1), parent_point))
child = RigidBody('child', child_point, child_frame, 1, (inertia(child_frame, 1, 1, 1), child_point))

# Set the joint axis
joint_axis = parent_frame.x

# Define the cylindrical joint
cylindrical_joint = CylindricalJoint('cylindricalJoint', parent, child,
                                     rotation_coordinate=q1,
                                     translation_coordinate=q2,
                                     rotation_speed=u1,
                                     translation_speed=u2,
                                     parent_point=parent_point,
                                     child_point=child_point,
                                     joint_axis=joint_axis)

# Define the wrapping geometry
cylinder = WrappingCylinder(r, O, joint_axis)

# Pathway
pathway = WrappingPathway(parent_point, child_point, geometry=cylinder)

# Normal force
normal_force = N * joint_axis

# Coulomb kinetic friction
friction = CoulombKineticFriction(mu_k, normal_force, pathway)

friction_force = friction.force
print("Friction Force: ", friction_force)

FL = [(child_point, friction_force)]
BL = [child]

coordinates = [q1, q2]
speeds = [u1, u2]

# Kane's method
kane = KanesMethod(parent_frame, coordinates, speeds, kd_eqs=[q1.diff(t) - u1, q2.diff(t) - u2])

fr, frstar = kane.kanes_equations(BL, FL)
```
In this scenario, `CoulombKineticFriction` works well with other pathways, i.e., WrappingPathway, and general friction cases. But the problem is that this change is not able to include the viscous term since the calculated force should be either a vector or a scalar (there's an error when introducing viscous term).

So I might suggest two options:
i. Use the modified force calculation without the viscous term.
ii. Use the current force calculation with the assumption that the normal force in this friction actuator is regarded as a positive scalar.

Reply:
Hi Hwayeon, you add this to the PR so that we can easily clone your branch and run the code?

Please use the `to_loads()` feature of the actuator to generate the forces for KanesMethod, LagrangesMethod, or System.

Reply:
`to_loads()` ensures that the signs of things are automatically correct.

Reply:
This is the more general cylindrical joint I was trying to describe:
![image](https://github.com/user-attachments/assets/580a412a-0b53-4580-86eb-e4a0bb093274)

Imagine a curve in 3D space and a cylinder can slide along the curve. The force that keeps the cylinder on the curve is the normal force and it depends on the speed and acceleration of the cylinder along the curve. This normal force is a noncontributing force which can be described by a vector quantity and it is a function of the motion of the cylinder as well as gravity (if present). A user can calculate this force using auxiliary speeds in Kane's method or lagrange multipliers in Lagrange's method and could provide it to a friction actuator for the correct kinetic friction to arise.

Reply:
To be clear, you don't have to make your actuator work for the most general models, but if it doesn't it needs to be documented to the user what the limitations are and where it can be utilized.

Reply:
I just made a new commit with two friction actuator classes: `CoulombKineticFriction` and `CoulombKineticFriction2`. Since there's an issue with the force expression in the friction actuator that treats the normal force as a vector, I retained the original class and created a new class temporarily.

`CoulombKineticFriction` is the original friction actuator that assumes the normal force is a positive scalar. I explained in the docstring which systems work well with this actuator and its limitations so that users can better understand. Also, the variable names are changed from `normal_force` and `viscous_coefficient` to `f_n` and `sigma` for consistency.

`CoulombKineticFriction2` is the new friction actuator that assumes the normal force is a vector. It does not include a viscous term since the calculated force should be either a scalar or vector. Variable names are also changed for consistency. The `to_loads` method is added to ensure compatibility with various pathways, including `WrappingPathway`. Now, there is a problem with the force expression.

In `TestCoulombKineticFriction`, only the variable names are changed.

`TestCoulombKineticFriction2` is a new test class for `CoulombKineticFriction2`. The tested system is the same as in `TestCoulombKineticFriction`. But there is a problem with force expression.

`TestCylinderFriction` is a new test class for `CoulombKineticFriction2`. It aims to test a cylinder sliding along a curve in 3D space you explained. Because I think the problem with the force expression in `CoulombKineticFriction2` and `TestCoulombKineticFriction2` should be addressed first, this class is currently incomplete.